### PR TITLE
Enhancement/refactoring

### PR DESCRIPTION
**Pull request by benjam-es**
Some streamlining, and extraction of trait and import code, making it usable for model generators. Further discussion on issue #527

### CODE REVIEW

**Comment by benjam-es** [comment]
Seems to be some conflict with the pull request from #531, not sure what that is yet. Will see if I can resolve.

**Comment by jasonmccreary** [comment]
@benjam-es, I believe you use some PHP 7.4+ only features. Hence 7.3 is failing.

I'll probably set a PHP 8.0 requirement later this month once Laravel 9 is released.

**Comment by benjam-es** [comment]
I see i'm still getting them... ill fire docker up with 7.3 and check it out rather than keep trying to push

File: src/Concerns/HandlesImports.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+trait HandlesImports
+{
+    protected function addImport(Model $model, $class)
+    {
+        $this->imports[$model->name()][] = $class;
+    }
+
+    protected function buildImports(Model $model)
+    {
+        return collect($this->imports[$model->name()])
```
Comment (jasonmccreary) [inline src/Concerns/HandlesImports.php:16]: Short closure is breaking 7.3 compatibility. Also `unique` and `sort` could be done before map for a minor optimization.

File: src/Generators/AbstractClassGenerator.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+{
+    use HandlesImports, HandlesTraits;
+
+    const INDENT = '        ';
+
+    protected Filesystem $filesystem;
+
+    private Tree $tree;
+
+    protected array $imports = [];
```
Comment (jasonmccreary) [inline src/Generators/AbstractClassGenerator.php:22]: Is it possible for these respective property to be added to the `trait` and not the abstract class?

File: src/Generators/ControllerGenerator.php [hunk 3]
```diff
@@ -17,63 +18,30 @@
...
         $stub = $this->filesystem->stub('controller.class.stub');
 
         /** @var \Blueprint\Models\Controller $controller */
         foreach ($tree->controllers() as $controller) {
-            $this->addImport($controller, 'Illuminate\\Http\\Request');
-
+            $this->addImport($controller, \Illuminate\Http\Request::class);
             if ($controller->fullyQualifiedNamespace() !== 'App\\Http\\Controllers') {
-                $this->addImport($controller, 'App\\Http\\Controllers\\Controller');
+                $this->addImport($controller, \App\Http\Controllers\Controller::class);
```
Comment (jasonmccreary) [inline src/Generators/ControllerGenerator.php:37]: Referencing this via its static `class` property may yield an error if it doesn't exist. Hence the original string method.

Also, aside from this PR, it should probably respect the configured namespaces.

File: tests/Feature/Lexers/SeederLexerTest.php [hunk 4]
```diff
@@ -57,7 +46,7 @@ public function it_returns_seeders()
         $this->assertIsArray($actual['seeders']);
         $this->assertCount(1, $actual['seeders']);
 
-        $this->assertSame(['Post'], $actual['seeders']);
+        $this->assertEquals(['Post'], $actual['seeders']);
```
Comment (jasonmccreary) [inline tests/Feature/Lexers/SeederLexerTest.php:49]: Why was there a need for the less strict assertion?

**Review by jasonmccreary — CHANGES_REQUESTED** [review]
I'm glad to merge this, provided two things are fixed.

1. Preserve PHP 7.3 compatibility (no short closures).
2. Lessen the generic use of properties and traits (e.g. move respective properties inside trait, only add trait to class that uses it)

**Comment by jasonmccreary** [comment]
@benjam-es, given this a review. Looks good. If you can make those final tweaks, awesome. If not, I'll work on it Tuesday.

File: tests/Feature/Lexers/SeederLexerTest.php (hunk 4 above)
Comment (benjam-es) [reply tests/Feature/Lexers/SeederLexerTest.php]: Believe this was left over from a different change I made and then removed.

File: src/Generators/AbstractClassGenerator.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+use Blueprint\Blueprint;
+use Blueprint\Concerns\HandlesImports;
+use Blueprint\Concerns\HandlesTraits;
+use Blueprint\Contracts\Model;
+use Illuminate\Filesystem\Filesystem;
+
+class AbstractClassGenerator
+{
+    use HandlesImports, HandlesTraits;
```
Comment (jasonmccreary) [inline src/Generators/AbstractClassGenerator.php:13]: Is it true that all subclasses handle imports? Same for traits?

File: src/Generators/AbstractClassGenerator.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    const INDENT = '        ';
+
+    protected $filesystem;
+
+    private $tree;
+
+    protected $imports = [];
+    protected $traits = [];
```
Comment (jasonmccreary) [inline src/Generators/AbstractClassGenerator.php:22]: Can these property not be added to their respective traits?

**Review by jasonmccreary — CHANGES_REQUESTED** [review]
Tests are passing, just the final questions on the genericism.

File: src/Generators/AbstractClassGenerator.php (hunk 5 above)
Comment (benjam-es) [reply src/Generators/AbstractClassGenerator.php:13]: It's not true that all subclasses handle imports and traits. It is currently Controller/Seeder/Factory/Test Generators (x4) that make use of the traits, and potentially the 5th with a view to the Model generator starting to use it. So with 5 classes, its around +20 lines (4 per class) and -4 lines from the abstract class.

I was going with the view that, all class generators should have the 'ability' to handle imports and traits, but I take your point, given that Migration, Route and Statement generators don't use this.

I have amended the pull request, along with the other property suggestions

**Comment by jasonmccreary** [comment]
@benjam-es, thanks! I think this is so much cleaner. 💅

**Pull request merged**